[clang][dataflow] Add a lattice to help cache const accessor methods

jvoung opened the pull request:
By caching const accessor methods we can sometimes treat method call results as stable (e.g., for issue https://github.com/llvm/llvm-project/issues/58510). Users can clear the cache when a non-const method is called that may modify the state of an object.
This is represented as mixin.

It  will be used in a follow on patch to change bugprone-unchecked-optional-access's lattice from NoopLattice to CachedConstAccessorsLattice<NoopLattice>, along with some additional transfer functions.

jvoung committed:
[clang][dataflow] Add a lattice to help represent cache const accessor methods

By caching const accessor methods we can sometimes treat method call
results as stable (e.g., for issue
https://github.com/llvm/llvm-project/issues/58510).
Users can clear the cache when a non-const method is called that may
modify the state of an object.

This is represented as mixin.
It  will be used in a follow on patch to change
bugprone-unchecked-optional-access's lattice from
NoopLattice to CachedConstAccessorsLattice<NoopLattice>,
along with some additional transfer functions.

jvoung committed:
clang-format

Xazax-hun commented on clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp:
```diff
@@ -0,0 +1,217 @@
+      : AST(R"cpp(
+    struct S {
+      int *valProperty() const;
+      int &refProperty() const;
```
I'd like to see a test with a method returning a record.

jvoung replied:
Thanks! Added a test "SameStructValBeforeClearOrDiffAfterClear"

Xazax-hun reviewed (commented):
Overall looks good to me but also wait for a review from the code owners.

ymand commented on clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h:
```diff
@@ -0,0 +1,218 @@
+  using Base::Base; // inherit all constructors
+
+  /// Creates or returns a previously created `Value` associated with a const
+  ///  method call `obj.getFoo()` where `RecordLoc` is the
```
```suggestion
  /// method call `obj.getFoo()` where `RecordLoc` is the
```

jvoung replied:
Fixed!

ymand commented on clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h:
```diff
@@ -0,0 +1,218 @@
+  getOrCreateConstMethodReturnValue(const RecordStorageLocation &RecordLoc,
+                                    const CallExpr *CE, Environment &Env);
+
+  /// Creates or returns a previously created `StorageLocation` associated a
```
```suggestion
  /// Creates or returns a previously created `StorageLocation` associated with a
```

jvoung replied:
Fixed!

ymand commented on clang/unittests/Analysis/FlowSensitive/CachedConstAccessorsLatticeTest.cpp:
```diff
@@ -0,0 +1,217 @@
+    void target() {
+      S s;
+      s.valProperty();
+      S s2;
```
Can we also test: 2 objects, same function called, distinct results? That would catch bad keying on the object (for example).

jvoung replied:
Thanks! Added "DifferentValsFromDifferentLocs"

ymand reviewed (approved):
modulo the requests for tests. This looks great!

jvoung committed:
Fix comments and add two more tests.

- a test for an accessor returning a record type
- a test for calling accessors of different objects

jvoung committed:
Fix formatting

jvoung committed:
Merge branch 'main' into unchecked_optional_cache_lattice

nico commented on clang/include/clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h:
```diff
@@ -0,0 +1,218 @@
+  ConstMethodReturnStorageLocationsType ConstMethodReturnStorageLocations;
+};
+
+namespace internal {
```
This breaks building on Windows (arguably by exposing an existing bug):

http://45.33.8.238/win/95986/step_3.txt (scroll to the very bottom):

```
../../clang/include\clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h(157,29): error: reference to 'internal' is ambiguous
  157 |   ConstMethodReturnValues = internal::joinConstMethodMap<Value>(
      |                             ^
../../clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp(1083,31): note: in instantiation of member function 'clang::dataflow::DataflowAnalysis<clang::dataflow::UncheckedOptionalAccessModel, clang::dataflow::CachedConstAccessorsLattice<clang::dataflow::NoopLattice>>::joinTypeErased' requested here
 1083 | UncheckedOptionalAccessModel::UncheckedOptionalAccessModel(ASTContext &Ctx,
      |                               ^
../../clang/include\clang/Analysis/FlowSensitive/CachedConstAccessorsLattice.h(109,11): note: candidate found by name lookup is 'clang::dataflow::internal'
  109 | namespace internal {
      |           ^
../../clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp(205,1): note: candidate found by name lookup is 'clang::dataflow::(anonymous namespace)::internal'
  205 | AST_MATCHER_P(CXXMemberCallExpr, publicReceiverType,
      | ^
../../clang/include\clang/ASTMatchers/ASTMatchersMacros.h(131,3): note: expanded from macro 'AST_MATCHER_P'
  131 |   AST_MATCHER_P_OVERLOAD(Type, DefineMatcher, ParamType, Param, 0)
      |   ^
../../clang/include\clang/ASTMatchers/ASTMatchersMacros.h(135,13): note: expanded from macro 'AST_MATCHER_P_OVERLOAD'
  135 |   namespace internal {                                                         \
      |             ^
```

Please take a look and revert for now if it takes a while to fix.

jvoung replied:
Ah -- Taking a look!

nico replied:
(Here's a build that needs less scrolling to find the diag: http://45.33.8.238/win/95994/step_3.txt)

jvoung replied:
Thanks! Created https://github.com/llvm/llvm-project/pull/113601

Though is there a way to validate this, similar to this clang-cl compile?
